File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.py
class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        m = defaultdict(int)
        unpaired = ans = 0
        for w in words:
            rev_w = w[::-1]
            if m[rev_w] > 0:
                ans += 4
                m[rev_w] -= 1
                if w[0] == w[1]:
                    unpaired -= 1
            else:
                m[w] += 1
                if w[0] == w[1]:
                    unpaired += 1
        if unpaired > 0:
            ans += 2
        return ans

File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.cpp
class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        unordered_map<string, int> same, diff;
        int res = 0, mxsame = 0;
        string mxstr;
        for(auto &w : words){
            if(w[0] == w[1]){
                same[w]++;
            } else {
                diff[w]++;
            }
        }
        for(auto &p : same){
            if(p.second%2 == 1 && mxsame < p.second){
                mxsame = p.second;
                mxstr = p.first;
            }
        }
        res += mxsame*2;
        for(auto &w : words){
            if(w[0] == w[1]) {
                if(w != mxstr){
                    res += 4*(same[w]/2);
                    same[w] = 0;
                }
                
                continue;
            }
            
            string opp = w;
            reverse(opp.begin(), opp.end());
            // cout << w << " " << opp << endl;
            int val = min(diff[w], diff[opp]);
            diff[w] -= val;
            diff[opp] -= val;
            res += 4*val;
        }
        
        return res;
    }
};